src/compile/selection/selection.ts

```diff
@@ -150,42 +151,41 @@ export function assembleTopLevelSignals(model: Model) {
 }
 
 export function assembleUnitData(model: UnitModel, data: VgData[]): VgData[] {
-  return data
-    .concat(Object.keys(model.component.selection)
-      .map(function(k: string) {
-        return {name: k + STORE};
-      }));
+  forEachSelection(model, selCmpt => {
+    data.push({name: selCmpt.name + STORE});
+  });
+
+  return data;
 }
 
 export function assembleUnitMarks(model: UnitModel, marks: any[]): any[] {
-  let clippedGroup = false,
+  let clipGroup = false,
       selMarks = marks;
-  forEachSelection(model, function(selCmpt, selCompiler) {
+  forEachSelection(model, (selCmpt, selCompiler) => {
     selMarks = selCompiler.marks ? selCompiler.marks(model, selCmpt, selMarks) : selMarks;
-    forEachTransform(selCmpt, function(txCompiler) {
-      clippedGroup = clippedGroup || txCompiler.clippedGroup;
+    forEachTransform(selCmpt, (txCompiler) => {
+      clipGroup = clipGroup || txCompiler.clipGroup;
       if (txCompiler.marks) {
         selMarks = txCompiler.marks(model, selCmpt, marks, selMarks);
       }
     });
   });
 
-  if (clippedGroup) {
-    selMarks = [{
-      type: 'group',
-      encode: {
-        enter: {
-          width: {field: {group: 'width'}},
-          height: {field: {group: 'height'}},
-          fill: {value: 'transparent'},
-          clip: {value: true}
-        }
-      },
-      marks: selMarks.map(model.correctDataNames)
-    }];
+  if (model.parent && model.parent instanceof LayerModel) {
+    return [selMarks, clippedGroup];
```

Comment:
I see that you later check `unit[1]` in `assembleLayerMarks` but not sure why.

Can you add a comment (either here or in `assembleLayerMarks` or both)  to explain why if the parent is a `LayerModel`, we have to set this up differently like this?